roaring/container_stash.go

```diff
@@ -35,28 +36,85 @@ const (
 // less than 4,096 values, an array is often used. Containers with long runs of
 // integers would use run length encoding, and more random data usually uses
 // bitmap encoding.
+//
+// the names "nnn" and "typp" are temporary hacks to make failure to update
+// code into a compiler error.
 type Container struct {
 	pointer  *uint16                  // the data pointer
 	len, cap int32                    // length and cap
-	n        int32                    // number of integers in container
-	mapped   bool                     // mapped directly to a byte slice when true
-	typ      byte                     // array, bitmap, or run
+	nnn      int32                    // number of integers in container
+	flags    containerFlags           // internal flags
+	typp     byte                     // array, bitmap, or run
 	data     [stashedArraySize]uint16 // immediate data for small arrays or runs
 }
 
+type containerFlags uint8
+
+const (
+	flagMapped = containerFlags(1 << iota)
+	flagFrozen
+)
+
+func (c *Container) String() string {
+	if c == nil {
+		return "<nil container>"
+	}
+	froze := ""
+	switch c.flags {
+	case flagFrozen:
+		froze = "frozen "
+	case flagMapped:
+		froze = "mapped "
+	case flagFrozen | flagMapped:
+		froze = "frozen/mapped"
+	}
+	switch c.typp {
+	case containerArray:
+		return fmt.Sprintf("<%sarray container, N=%d>", froze, c.N())
+	case containerBitmap:
+		return fmt.Sprintf("<%sbitmap container, N=%d, len %dx uint64>",
+			froze, c.N(), len(c.bitmap()))
+	case containerRun:
+		return fmt.Sprintf("<%srun container, N=%d, len %dx interval>",
+			froze, c.N(), len(c.runs()))
+	default:
+		return fmt.Sprintf("<unknown %s%d container, N=%d>", froze, c.typp, c.N())
+	}
+}
+
 // NewContainer returns a new instance of container. This trivial function
 // may later become more interesting.
 func NewContainer() *Container {
 	statsHit("NewContainer")
-	c := &Container{typ: containerArray, len: 0, cap: stashedArraySize}
+	c := &Container{typp: containerArray, len: 0, cap: stashedArraySize}
 	c.pointer = (*uint16)(unsafe.Pointer(&c.data[0]))
 	return c
 }
 
 // NewContainerBitmap makes a bitmap container using the provided bitmap, or
 // an empty one if provided bitmap is nil. If the provided bitmap is too short,
 // it will be padded.
-func NewContainerBitmap(n int32, bitmap []uint64) *Container {
+func NewContainerBitmap(bitmap []uint64) *Container {
```

Comment:
we can't change this signature without a 2.0 release

Reply:
Yeah. I could in theory not change the NewContainerBitmap calls, but since we need the changes to other things (e.g., optimize, Update), we probably need the 2.0 API change.

Reply:
Updated to note: We don't need the API change! Update can just refuse to run on frozen things, possibly panicing if roaringParanoia is on, but also we can just not call it in the case that's affected, and instead make a new function that does what we need. Also reverted this change, and I think I'm back to no API breakage, although I'm not totally sure.